File: qt-src/interface/buttoncell.cpp
#include "buttoncell.h"

ButtonCell::ButtonCell(Language* language, QToolButton *parent) :
    QToolButton(parent),
    language(language)
{
    connect(this, &ButtonCell::clicked, this, &ButtonCell::onLetterClicked);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
    createMenu();
}

void ButtonCell::connectToPanel(QObject *gb)
{
    connect(this, SIGNAL(onLetterChosen(QChar)), gb, SLOT(onCellChosen(QChar)));
    connect(this, SIGNAL(onLetterPushed()), gb, SLOT(onCellPushed()));
}

void ButtonCell::createMenu() {
    signalMapper_ = new QSignalMapper(this);
    QMenu* menu = new QMenu(tr("Set of letters"));
    QVector <QChar> currentLang = language->getList();
    for(int i = 0; i<currentLang.size(); i++){
        actions_.push_back(new QAction(QString(currentLang[i]), this));
        menu->addAction(actions_.back());
        signalMapper_->setMapping(actions_.back(), QString(currentLang[i]));
        connect(actions_.back(), SIGNAL(triggered()), signalMapper_, SLOT(map()));
    }
    connect(signalMapper_, SIGNAL(mapped(QString)), this, SLOT(setLetter(QString)));
    setMenu(menu);
}

File: balda_game/src/interface/ButtonCell.py
__author__ = 'akhtyamovpavel'

from PySide import QtCore, QtGui

class ButtonCell(QtGui.QToolButton):
    on_letter_chosen = QtCore.Signal(str)
    on_letter_pushed = QtCore.Signal()


    @QtCore.Slot(str)
    def set_letter(self, letter):
        self.on_letter_chosen.emit(letter)

    @QtCore.Slot()
    def on_letter_clicked(self):
        self.on_letter_pushed.emit()

    def __init__(self, language):
        super(ButtonCell, self).__init__()
        self.__language__ = language
        self.clicked.connect(self.on_letter_clicked)

        self.setSizePolicy(QtGui.QSizePolicy.Expanding, QtGui.QSizePolicy.Preferred)
        self.create_menu()


    def create_menu(self):
        self.signal_mapper = QtCore.QSignalMapper(self)
        self.__menu__ = QtGui.QMenu('Set of letters')

        self.__actions__ = list()

        current_language = self.__language__.get_list()

        for letter in current_language:
            action = QtGui.QAction(letter, self)


            self.__menu__.addAction(action)
            self.signal_mapper.setMapping(action, letter)

            action.triggered.connect(self.signal_mapper.map)
            self.__actions__.append(action)

        self.signal_mapper.mapped[str].connect(self.set_letter)
        self.setMenu(self.__menu__)

    def connect_to_panel(self, gb):
        self.on_letter_chosen.connect(gb.on_cell_chosen)
        self.on_letter_pushed.connect(gb.on_cell_pushed)

